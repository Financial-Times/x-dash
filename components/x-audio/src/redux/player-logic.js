// intial state
export const initialState = {
	playing: false,
	loading: false,
	lastActionInternal: true
}

// reducer
export function reducer (state = initialState, action) {
	switch (action.type) {
		case 'PLAY':
			return {...state, playing: true }
		case 'PAUSE':
			return {...state, playing: false }
		case 'REQUEST_PLAY':
		case 'REQUEST_PAUSE':
			return {...state, lastActionInternal: action.isInternal }
		default:
			return state;

	}
}


// actions
export const actions = {
	play: () => ({
		type: 'PLAY'
	}),
	pause: () => ({
		type: 'PAUSE'
	}),
	requestPlay: ({ url, isInternal = true } = {}) => ({
		type: 'REQUEST_PLAY',
		url,
		isInternal
	}),
	requestPause: ({ isInternal = true } = {}) => {
		console.log({
			type: 'REQUEST_PAUSE',
			isInternal
		})
		return ({
			type: 'REQUEST_PAUSE',
			isInternal
		})
	},
}


// middleware
export const middleware = store => {

	const audio = new Audio();
	// debuging
	[
		'loadeddata',
		'loadedmetadata',
		'loadstart',
		'progress',
		'canplay',
		'canplaythrough'
	].forEach(evtName => {
		audio.addEventListener(evtName, () => {
			console.log('audio eventy', evtName);
		});
	});

	audio.addEventListener('play', () => store.dispatch(actions.play()));

	audio.addEventListener('pause', () => store.dispatch(actions.pause()));

	return next => action => {
		switch (action.type) {
			case 'REQUEST_PLAY':
				if (action.url) audio.src = action.url;
				audio.play();
				break;
			case 'REQUEST_PAUSE':
				audio.pause();
				break;
		}
		next(action);
	}
}

// debug
