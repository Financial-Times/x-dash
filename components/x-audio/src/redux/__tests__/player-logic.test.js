import { middleware, actions, initialState, reducer } from '../player-logic';
const runActions = (initialState, ...actions) => actions.reduce(reducer, initialState);

describe('actions and reducer', () => {

	test('Actions not relevant to audio have no effect on state', () => {
		const updatedState = runActions(initialState, { type: 'UNKNOWN_ACTION' });

		expect(updatedState).toBe(initialState);
	});

	test('Play action sets playing to true', () => {
		const updatedState = runActions(initialState, actions.play());

		expect(updatedState).toMatchSnapshot();
	});

	test('Pause action sets playing to false', () => {
		const updatedState = runActions(initialState, actions.play(), actions.pause());

		expect(updatedState).toMatchSnapshot();
	});

	test('Loading action sets loading to true', () => {
		const updatedState = runActions(initialState, actions.loading());

		expect(updatedState).toMatchSnapshot();
	});

	test('Loaded action sets loading to false', () => {
		const updatedState = runActions(initialState, actions.loading(), actions.loaded());

		expect(updatedState).toMatchSnapshot();
	});

	test('Update current time action updates currentTime', () => {
		const currentTime = 10;
		const updatedState = runActions(initialState, actions.updateCurrentTime({currentTime}));

		expect(updatedState).toMatchSnapshot();
	});

});

describe('middleware', () => {

	const create = (state = initialState) => {
		const store = {
			getState: jest.fn(() => state),
			dispatch: jest.fn()
		}
		const next = jest.fn();
		const audio = new Audio();
		audio.play = jest.fn();
		audio.pause = jest.fn();
		const middlewareWithNext = middleware(store, audio)(next);
		const invoke = action => middlewareWithNext(action)

		return { store, next, invoke, audio };
	}

	test('requestPlay with url sets src and plays', () => {
		const { invoke, audio } = create();
		invoke(actions.requestPlay({ url: 'https://local.ft.com/url' }));

		expect(audio.play).toHaveBeenCalled();
		expect(audio.src).toBe('https://local.ft.com/url');
	});

	test('requestPlay without url just plays', () => {
		const { invoke, audio } = create();
		const url = 'https://local.ft.com/url';
		audio.src = url;
		invoke(actions.requestPlay({}));

		expect(audio.play).toHaveBeenCalled();
		expect(audio.src).toBe(url);
	});

	test('requestPause pauses', () => {
		const { invoke, audio } = create();
		invoke(actions.requestPause());

		expect(audio.pause).toHaveBeenCalled();
	});

	test('HTML play event dispatches play action', () => {
		const { store, audio } = create();
		audio.dispatchEvent(new Event('play'));

		expect(store.dispatch).toHaveBeenCalledWith(actions.play());
	});

	test('HTML pause event dispatches pause action', () => {
		const { store, audio } = create();
		audio.dispatchEvent(new Event('pause'));

		expect(store.dispatch).toHaveBeenCalledWith(actions.pause());
	});

	[
		'waiting',
		'stalled',
		'loadstart',
		'loadedmetadata',
		'loadeddata'
	].forEach(action => {
		test(`HTML ${action} event dispatches loading action`, () => {
			const { store, audio } = create();
			audio.dispatchEvent(new Event(action));

			expect(store.dispatch).toHaveBeenCalledWith(actions.loading());
		});
	})

	test('HTML canplay event dispatches loaded action', () => {
		const { store, audio } = create();
		audio.dispatchEvent(new Event('canplay'));

		expect(store.dispatch).toHaveBeenCalledWith(actions.loaded());
	});

	test('HTML timeupdate event dispatches updateCurrentTime action', () => {
		const { store, audio } = create({ currentTime: 9 });
		audio.currentTime = 10;
		audio.dispatchEvent(new Event('timeupdate'));

		expect(store.dispatch).toHaveBeenCalledWith(actions.updateCurrentTime({currentTime: 10}));
	});

	test('updateCurrentTime action is only dispatched once per second', () => {
		const { store, audio } = create({ currentTime: 10 });
		audio.currentTime = 10.5;
		audio.dispatchEvent(new Event('timeupdate'));

		expect(store.dispatch).not.toHaveBeenCalled();
	});

});
