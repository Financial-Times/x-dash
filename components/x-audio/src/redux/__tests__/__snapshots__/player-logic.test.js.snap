// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions and reducer Ended action sets ended to true 1`] = `
Object {
  "currentTime": 0,
  "ended": true,
  "loading": false,
  "playing": true,
}
`;

exports[`actions and reducer Loaded action sets loading to false 1`] = `
Object {
  "currentTime": 0,
  "ended": false,
  "loading": false,
  "playing": false,
}
`;

exports[`actions and reducer Loading action sets loading to true 1`] = `
Object {
  "currentTime": 0,
  "ended": false,
  "loading": true,
  "playing": false,
}
`;

exports[`actions and reducer Pause action sets playing to false 1`] = `
Object {
  "currentTime": 0,
  "ended": false,
  "loading": false,
  "playing": false,
}
`;

exports[`actions and reducer Play action sets playing to true 1`] = `
Object {
  "currentTime": 0,
  "ended": false,
  "loading": false,
  "playing": true,
}
`;

exports[`actions and reducer Playing audio after previous has ended resets ended state 1`] = `
Object {
  "currentTime": 0,
  "ended": false,
  "loading": false,
  "playing": true,
}
`;

exports[`actions and reducer Requesting new audio play resets ended state 1`] = `
Object {
  "currentTime": 0,
  "ended": false,
  "loading": false,
  "playing": false,
}
`;

exports[`actions and reducer Update current time action updates currentTime 1`] = `
Object {
  "currentTime": 10,
  "ended": false,
  "loading": false,
  "playing": false,
}
`;
